File: Olympiad/COCI/official/2008/contest2_solutions/turbo.pas
program turbo;

const MAX = 100000;

var
   i, n : longint;
   mini, maxi : longint;

   numbers : array[1..MAX] of longint;
   position : array[1..MAX] of longint;
   fenwick : array[1..MAX] of longint;

function query( x : longint ) : longint;
var
   ret : longint;
begin
   ret := 0;
   while x > 0 do begin
      ret := ret + fenwick[x];
      x := x - (x and -x);
   end;
   query := ret;
end;

function sum( lo, hi : longint ) : longint;
begin
   sum := query( hi ) - query( lo-1 );
end;

procedure add( x, val : longint );
begin
   while x <= n do begin
      fenwick[x] := fenwick[x] + val;
      x := x + (x and -x);
   end;
end;

begin
   readln( n );
   for i := 1 to n do begin
      readln( numbers[i] );
      position[numbers[i]] := i;
      add( i, 1 );
   end;

   mini := 1;
   maxi := n;

   for i := 1 to n do begin
      if i mod 2 = 1 then begin

         add( position[mini], -1 );
         writeln( sum( 1, position[mini] ) );
         mini := mini + 1;

      end else begin

         add( position[maxi], -1 );
         writeln( sum( position[maxi], n ) );
         maxi := maxi - 1;

      end;
   end;
end.

File: Olympiad/COCI/official/2008/contest2_solutions/turbo.cpp
#include <cstdio>
#include <cstring>

#define MAX 100000

int n;
int numbers[MAX+1];
int position[MAX+1];

struct fenwick {
   int a[MAX+1];

   fenwick() {
      memset( a, 0, sizeof a );
   }

   int query( int X ) {
      int ret = 0;
      for( int x = X; x > 0; x -= x&-x )
         ret += a[x];
      return ret;
   }

   int sum( int lo, int hi ) {
      return query( hi ) - query( lo-1 );
   }
   
   void add( int X, int val ) {
      for( int x = X; x <= n; x += x&-x )
         a[x] += val;
   }
} alive;

int main( void ) {
   
   scanf( "%d", &n );
   for( int i = 1; i <= n; ++i ) {
      scanf( "%d", &numbers[i] );
      position[numbers[i]] = i;
      alive.add( i, 1 );
   }

   int mini = 1, maxi = n;

   for( int i = 1; i <= n; ++i ) {
      if( i%2 == 1 ) {
         alive.add( position[mini], -1 );
         printf( "%d\n", alive.sum( 1, position[mini] ) );
         ++mini;
      } else {
         alive.add( position[maxi], -1 );
         printf( "%d\n", alive.sum( position[maxi], n ) );
         --maxi;
      }
   }

   return 0;
}
